jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c

```diff
@@ -1657,286 +1528,346 @@ ecma_builtin_string_prototype_object_split (ecma_value_t this_arg, /**< this arg
+
+        /* 6. */
+        const ecma_length_t string_length = ecma_string_get_length (this_to_string_p);
+
+        while (curr_pos < string_length && !should_return && ecma_is_value_empty (ret_value))
         {
-          /* 11.a */
-          ecma_value_t match_result = ecma_builtin_helper_split_match (this_to_string_val,
-                                                                       0,
-                                                                       separator);
+          ecma_value_t match_result = ecma_make_simple_value (ECMA_SIMPLE_VALUE_NULL);
 
-          /* 11.b */
-          if (!ecma_is_value_null (match_result))
+          if (separator_is_regexp)
           {
-            should_return = true;
+#ifndef CONFIG_DISABLE_REGEXP_BUILTIN
+            ecma_value_t regexp_value = ecma_copy_value_if_not_object (separator);
+            ecma_string_t *substr_str_p = ecma_string_substr (this_to_string_p, curr_pos, string_length);
+            match_result = ecma_regexp_exec_helper (regexp_value, ecma_make_string_value (substr_str_p), true);
+            ecma_deref_ecma_string (substr_str_p);
+#else
+            return ecma_raise_type_error (ECMA_ERR_MSG ("REGEXP separator is disabled in split method."));
+#endif
           }
           else
           {
-            /* 11.c */
-            ecma_string_t *zero_str_p = ecma_new_ecma_string_from_number (ECMA_NUMBER_ZERO);
+            ecma_string_t *separator_str_p = ecma_get_string_from_value (separator);
+            ecma_length_t separator_length = ecma_string_get_length (separator_str_p);
 
-            ecma_value_t put_comp = ecma_builtin_helper_def_prop (new_array_p,
-                                                                  zero_str_p,
-                                                                  this_to_string_val,
-                                                                  true,
-                                                                  true,
-                                                                  true,
-                                                                  false);
-
-            JERRY_ASSERT (ecma_is_value_true (put_comp));
-
-            /* 11.d */
-            should_return = true;
+            if (curr_pos + separator_length <= string_length)
+            {
+              bool is_different = false;
+              for (ecma_length_t i = 0; i < separator_length && !is_different; i++)
+              {
+                ecma_char_t char_from_string = ecma_string_get_char_at_pos (this_to_string_p, curr_pos + i);
+                ecma_char_t char_from_separator = ecma_string_get_char_at_pos (separator_str_p, i);
+                is_different = (char_from_string != char_from_separator);
+              }
 
-            ecma_deref_ecma_string (zero_str_p);
+              if (!is_different)
+              {
+                /* 6-7. */
+                match_result = ecma_op_create_array_object (0, 0, false);
+              }
```

Comment:
Yes, the for loop must end if the `char_from_string != char_from_separator` condition is true, but we should call the array object creation if that condition is false after the for loop.